drivers/wifi/esp32/src/esp_wifi_drv.c

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2020 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_esp32_wifi
+
+#define LOG_MODULE_NAME esp32_wifi
+#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <net/ethernet.h>
+#include <net/net_pkt.h>
+#include <net/net_if.h>
+#include <device.h>
+#include <soc.h>
+#include <ethernet/eth_stats.h>
+#include "esp_networking_priv.h"
+#include "esp_private/wifi.h"
+#include "esp_event.h"
+#include "esp_timer.h"
+#include "esp_wifi_system.h"
+#include "esp_wpa.h"
+
+#define DEV_DATA(dev) \
+	((struct esp32_wifi_runtime *)(dev)->data)
+
+struct esp32_wifi_runtime {
+	struct net_if *iface;
+	uint8_t mac_addr[6];
+	bool tx_err;
+	uint32_t tx_word;
+	int tx_pos;
+#if defined(CONFIG_NET_STATISTICS_ETHERNET)
+	struct net_stats_eth stats;
+#endif
+};
+
+static int eth_esp32_send(const struct device *dev, struct net_pkt *pkt)
+{
+	const int pkt_len = net_pkt_get_len(pkt);
+	uint8_t *tx_buffer = (uint8_t *) k_malloc(pkt_len);
+
+	if (tx_buffer == NULL) {
+		return -EIO;
+	}
+
+	/* Read the packet payload */
+	if (net_pkt_read(pkt, tx_buffer, pkt_len) < 0) {
+		k_free(tx_buffer);
+		return -EIO;
+	}
+
+	/* Enqueue packet for transmission */
+	esp_wifi_internal_tx(ESP_IF_WIFI_STA, (void *)tx_buffer, pkt_len);
+	k_yield();
```

Comment:
Is this yield needed?
The documentation of `esp_wifi_internal_tx()` says that
`This API makes a copy of the input buffer and then forwards the buffer copy to WiFi driver.`
There is lot of copying here and then again in the internal tx func, I wonder would there be a better way for this. Could the HAL provide a buffer and we could copy the data directly to it. Then we would not need to call malloc either.

Reply:
I removed the `k_yield()` and added a static buffer so avoid copying data.